File: NetworkGame.py
#!/usr/bin/env python3
"""
Evolution of interacting networks.

This implements the evolution of the sequentially processing network of Le Nagard, Chao and Tenaillon, 2011, BMC Evolutionary Biology, for two interacting networks that together have to match an environment with varying complexity, represented by a Legendre Polynomial of varying order.

Erol Akcay <eakcay@sas.upenn.edu>
Jeremy Van Cleve <vancleve@santafe.edu>
"""

import sys
from argparse import ArgumentParser
from numpy import *
import h5py
from datetime import datetime
from scipy.stats import mode
# random.seed(100)


## define the function to generate the output of node j given previous outputs and weights and Wb
def calcOj (j, prev_out, Wm, Wb):
    x = dot(Wm[0:j,j],prev_out[0:j]) + Wb[j]
    return 1-exp(-x*x)

## iterate over the whole network
def iterateNetwork (input, Wm, Wb):
    prev_out=zeros(Wb.shape[0])
    prev_out[0]=input
    for j in range(1, Wb.shape[0]):
        prev_out[j]=calcOj(j, prev_out, Wm, Wb)
    return prev_out

## one round of two-player game
def networkGameRound (mutWm, mutWb, mutPrev, resWm, resWb, resPrev):
    mutOut = iterateNetwork(resPrev, mutWm, mutWb)[-1]
    resOut = iterateNetwork(mutPrev, resWm, resWb)[-1]

    return [mutOut, resOut]


## play `rounds` of the two-player game
def repeatedNetworkGame (rounds, mutWm, mutWb, mutInit, resWm, resWb, resInit):
    for i in range(rounds):
        [mutInit, resInit] = networkGameRound(mutWm, mutWb, mutInit, resWm, resWb, resInit)

    return [mutInit, resInit]

## play rounds but now save the entire history
def repeatedNetworkGameEntireHist (rounds, mutWm, mutWb, mutInit, resWm, resWb, resInit):
    muthist = zeros(rounds)
    resthist = zeros(rounds)
    for i in range(rounds):
        [mutInit, resInit] = networkGameRound(mutWm, mutWb, mutInit, resWm, resWb, resInit)
        muthist[i] = mutInit
        resthist[i] = resInit
        
    return [muthist, resthist]

####################################################################
## calulate resident-mutant fitness matrix from contributions at end of repeated cooperative
## investment game
##
## fitness = 1 + b * (partner) - c * (self) + d * (self) * (partner)
####################################################################

def fitnessOutcome (b, c, d, fitness_benefit_scale, rounds, mutWm, mutWb, mutInit, resWm, resWb, resInit):
    [rmOut, mrOut] = repeatedNetworkGame(rounds, resWm, resWb, resInit, mutWm, mutWb, mutInit)

    wmr = 1 + ((b * rmOut - c * mrOut + d * rmOut*mrOut)*fitness_benefit_scale)
    wrm = 1 + ((b * mrOut - c * rmOut + d * rmOut*mrOut)*fitness_benefit_scale)
   
    return [[wmr, wrm], [rmOut, mrOut]]

##################################################################
## calulate resident-mutant fitness matrix from contributions throughout the game history,
## discounted at rate delta going from present to the past in the repeated cooperative
## investment game
##
## fitness = 1 + b * (partner) - c * (self) + d * (self) * (partner)
##################################################################
def fitnessOutcomeEntireHist (b, c, d, fitness_benefit_scale, rounds, mutWm, mutWb, mutInit, resWm, resWb, resInit, delta):
    [rmOut, mrOut] = repeatedNetworkGameEntireHist(rounds, resWm, resWb, resInit, mutWm, mutWb, mutInit)
    discount = exp(-delta*(rounds-1-arange(0,rounds)))
    td = discount.sum()

    wmr = 1 + ((dot((b * rmOut - c * mrOut + d * rmOut*mrOut), discount))*fitness_benefit_scale)
    wrm = 1 + ((dot((b * mrOut - c * rmOut + d * rmOut*mrOut), discount))*fitness_benefit_scale)

    return [[wmr, wrm], [dot(rmOut, discount)/td, dot(mrOut, discount)/td]]


##################################################################
# Calculates fitness from games played, stores the value of each pair,
# and tracks the maximum observed fitness for each matchup type at,
# each timepoint.
##################################################################

def pairwise_fitness(population_array, repro_probabilities, genotypes_dict, fit_dict, fitFunc, w_max, init_max):
    shuffled_population = population_array.copy()
    random.shuffle(shuffled_population)
    mean_init = zeros((2*len(population_array)))
    payoffs = zeros(2*len(population_array))
    for n1, n2, site in zip(population_array, shuffled_population, range(len(population_array))):
        if n1 not in fit_dict.keys():
            fit_dict[n1] = {}
        if n2 not in fit_dict[n1].keys():
            
            fit_dict[n1][n2] = fitFunc(genotypes_dict[n2][0], genotypes_dict[n2][1], genotypes_dict[n2][2], genotypes_dict[n1][0], genotypes_dict[n1][1], genotypes_dict[n1][2])
            
        p = fit_dict[n1][n2][0]
        
        for w_i in range(2):
            if p[w_i] > w_max[w_i]:
                w_max[w_i] = p[w_i]
        
        if max([genotypes_dict[n1][2], genotypes_dict[n2][2]]) > init_max:
            init_max = max([genotypes_dict[n1][2], genotypes_dict[n2][2]])
        mean_init[site] = genotypes_dict[n1][2]
        mean_init[site+len(population_array)] = genotypes_dict[n2][2]
        payoffs[site] = fit_dict[n1][n2][1][0]
        payoffs[site+len(population_array)] = fit_dict[n1][n2][1][1]
        repro_probabilities[site] = p[0]

    return repro_probabilities, fit_dict, w_max, init_max, mean(mean_init), mean(payoffs)

##################################################################
#   Mutation function, samples each child in the offspring generation,
#   and then applies mutation to nodes, edges, and init as below
##################################################################

def mutation_process(population_size, genotypes_dict, n_genotypes, new_pop_array, matDim, mutlink, mutsize, mutinitsize, mu):
    for N in range(population_size):
        if random.rand() < mu:
            mutationw = triu(random.binomial(1,mutlink,(matDim,matDim))
                      *(random.normal(0,mutsize, (matDim,matDim))))
            mutWm = genotypes_dict[new_pop_array[N]][0] + mutationw
            mutationb = random.binomial(1,mutlink,matDim)*random.normal(0, mutsize, matDim)
            mutWb = genotypes_dict[new_pop_array[N]][1] + mutationb
            mutation_init = random.normal(0, scale = mutinitsize)
            mutInit = genotypes_dict[new_pop_array[N]][2] + mutation_init
            new_pop_array[N] = n_genotypes
            genotypes_dict[n_genotypes] = [mutWm,mutWb,mutInit]
            n_genotypes += 1
            
    return genotypes_dict, n_genotypes, new_pop_array

##################################################################
# the main simulation code, iterating the sequential mutation and invasion process
##################################################################

def simulation (initWm, initWb, initIn, population_size, mu, b, c, d, r, rounds, Tmax, mutsize, mutinitsize, mutlink, fitFunc):
    """
    initWm are the initial network weights
    initWb are the initial node weights
    b is the benefit
    c is the cost
    d is the synergy
    r is the relatedness
    rounds is the number of rounds of game play
    Tmax is the maximum number of iterations
    mutsize: 2*maximum mutation magnitude
    mutlink: probability that an entry will mutate
    """
    matDim=len(initWm)
    resWm = initWm
    resWb = initWb
    histlen = Tmax+1
    invashist = zeros(histlen, dtype=int64)
    fithistory = zeros(histlen)
    wmhist = zeros((histlen,) + resWm.shape)
    bhist = zeros((histlen,) + resWb.shape)
    nethist = zeros((histlen,2))

    # save initial values
    w, nout = fitFunc(resWm, resWb, initIn, resWm, resWb, initIn)
    fithistory[0] = w[0]
    wmhist[0] = resWm
    bhist[0] = resWb
    nethist[0] = nout
    ninvas = 0
    
    ################################################################
    #   Initializing population arrays, creating dict of genotypes,
    #   array of genotype indices, dicts for caching game results,
    #   and arrays for tracking history
    ################################################################
    
    genotypes_dict = {int(0) : [resWm, resWb, initIn]}
    population_array = zeros([Tmax, population_size], dtype = int)
    n_genotypes = 1
    fit_dict = {}
    previous_resident = int32(0)
    w_max_history = zeros((Tmax, 4))
    init_max_history = zeros(Tmax)
    mean_fitness_history = zeros(Tmax)
    mean_init_history = zeros(Tmax)
    mean_payoff_history = zeros(Tmax)
    
    
    ######################
    #     Time Start     #
    ######################
    
    for i in range(Tmax):
        
        ######################
        #  per time counters #
        ######################
        w_max = [0,0,0,0]
        init_max = float(0)
        repro_probabilities = zeros(population_size)
        current_population_array = population_array[i].copy()

        ######################
        #    reproduction    #
        ######################

        repro_probabilities, fit_dict, w_max, init_max, mean_init, mean_payoff = pairwise_fitness(current_population_array, repro_probabilities, genotypes_dict, fit_dict, fitFunc, w_max, init_max)
        new_pop_array = random.choice(current_population_array, 
                                      size = shape(current_population_array), 
                                      p = (repro_probabilities/sum(repro_probabilities))).tolist()

        ######################
        #    mutation code   #
        ######################
        
        genotypes_dict, n_genotypes, new_pop_array = mutation_process(population_size, genotypes_dict, n_genotypes, new_pop_array, matDim, mutlink, mutsize, mutinitsize, mu)
            
        
        ######################################################
        #   Updating output arrays. Resident is just         #
        #   whichever genotype has the largest frequency     #
        #   in the population. If more than mode exists,     #
        #   it picks the oldest genotype.                    #
        ######################################################

        current_resident = mode(new_pop_array)[0][0]
        if current_resident != previous_resident:
            wmhist[ninvas] = genotypes_dict[current_resident][0]
            bhist[ninvas] = genotypes_dict[current_resident][1]
            fithistory[ninvas] = fit_dict[current_resident][current_resident][0][0]
            ninvas += 1
            previous_resident = current_resident


        ##############################
        # update simulation history  #
        ##############################
        w_max_history[i] = w_max
        init_max_history[i] = init_max
        mean_fitness_history[i] = mean(repro_probabilities)
        mean_init_history[i] = mean_init
        mean_payoff_history[i] = mean_payoff
        if i != Tmax-1:
            population_array[i+1] = new_pop_array.copy()
        print('net out (Δmm, Δmr, rr)  : ({:.5f}, {:.5f}, {:.5f})'.format(nout[0]-nout[2], nout[1]-nout[2], nout[2])) if verbose == 2 else None

    return {'n_invas' : ninvas,
            'n_mutants' : n_genotypes,
            'w_max_hist' : w_max_history,
            'init_max_hist' : init_max_history,
            'mean_fitness_hist' : mean_fitness_history,
            'mean_init_hist' : mean_init_history,
            'timesteps' : Tmax,
            'invas_hist' : range(Tmax),
            'fit_hist' : fithistory[0:i],
            'wm_hist' : wmhist[0:i],
            'b_hist' : bhist[0:i]}


# Plot network with igraph
# def plotNetwork(wm):
#     """
#     Plot network using igraph package

#     wm: weight matrix
#     """
#     import igraph as ig

#     # create graph of only upper triangle
#     g = ig.Graph.Weighted_Adjacency(triu(wm,1).tolist())

#     # change edge widths and vertex colors
#     g.es["width"] = ig.rescale(g.es["weight"], out_range=(0.0, 4))
#     g.vs['color'] = 'blue'
#     g.vs[0]['color'] = 'green'
#     g.vs[g.vs.indices[-1]]['color'] = 'red'

#     return ig.plot(g, "output.png")


## Main function to run simulation
def main():
    parser = ArgumentParser(prog='command', description='Evolution of interacting networks')

    pars = ['nreps', 'tmax', 'rounds', 'population_size', 'mu', 'fitness_benefit_scale', 'b', 'c', 'd', 'r', 'nnet', 'initIn', 'initstddev', 'mutsize', 'mutinitsize', 'mutlink',
            'discount', 'seed', 'outputfile']

    parsdefault = dict(zip(pars,
                           [1, 1000, 10, 100, 0.01, 0.5, 2, 1, 0, 0 , 5, 0.1, 1, 0.1, 0.01, 0.5,
                            0.9, 0, 'output.h5']))
    
    parstype    = dict(zip(pars,
                           [int, int, int, int, float, float, float, float, float, float, int, float, float, float, float, float, float, int, str]))

    parshelp    = dict(zip(pars,
                           ['Number of times to replicate simulation (default: %(default)s)',
                            'Number of time steps (invasions) (default: %(default)s)',
                            'Number of rounds of game play (default: %(default)s)',
                            'Population Size for WF selection',
                            'Probability of mutation per birth',
                            'Multiplier of the the payoff calculated  per game round',
                            'payoff benefit (default: %(default)s)',
                            'payoff cost (default: %(default)s)',
                            'payoff synergy (default: %(default)s)',
                            'relatedness coefficient (default: %(default)s)',
                            'Number of nodes in the network (default: %(default)s)',
                            'Initial network input (game contribution) (default: %(default)s)',
                            'Initial weight matrix std dev (default: %(default)s)',
                            'Size of network mutational steps (std dev) (default: %(default)s)',
                            'Size of initial offer mutational steps (std dev) (default: %(default)s)',
                            'Probability weight mutates (default: %(default)s)',
                            'Payoff discount rate (negative value = use last round) (default: %(default)s)',
                            'seed for random number generator; if set to zero, use system time (default: %(default)s)',
                            'output file name for .h5 file (default: %(default)s)']))

    for par in pars:
        parser.add_argument("--"+par, default=parsdefault[par],
                            type=parstype[par], help=parshelp[par])

    parser.add_argument('-v', '--verbose', action="count",
                        help="increase output verbosity")

    args = parser.parse_args()

    global verbose
    verbose = args.verbose

    # set random seed specifically if given (set to system clock otherwise)
    if (args.seed):
        random.seed(args.seed)

    data = dict()
    for t in range(args.nreps):
        print('{}: running replicate {}'.format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), t+1))
        sys.stdout.flush() # flush stdout to make sure the print statement

        initWm = random.normal(0, args.initstddev, (args.nnet, args.nnet))
        initWb = random.normal(0, args.initstddev, (args.nnet))
        if args.discount >= 0:
            fitFunc = lambda mutWm, mutWb, mutInit, resWm, resWb, resInit: fitnessOutcomeEntireHist(args.b, args.c, args.d, args.fitness_benefit_scale, args.rounds, mutWm, mutWb, mutInit, resWm, resWb, resInit, args.discount)
        else:
            fitFunc = lambda mutWm, mutWb, mutInit, resWm, resWb, resInit: fitnessOutcome(args.b, args.c, args.d, args.fitness_benefit_scale, args.rounds, mutWm, mutWb, mutInit, resWm, resWb, resInit)
        simoutput = simulation(initWm, initWb, args.initIn, args.population_size, args.mu, args.b, args.c, args.d, args.r, args.rounds, args.tmax, args.mutsize, args.mutinitsize, args.mutlink, fitFunc)

        for key in simoutput.keys():
            if key in data:
                data[key] = concatenate((data[key], array(simoutput[key], ndmin=1)))
            else:
                data[key] = array(simoutput[key], ndmin=1)

        print('{}: replicate {} done'.format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), t+1))
        print("Number of Invasions: ", simoutput['n_invas'])
        print("Number of mutants: ", simoutput['n_mutants'])
#        outputdump = simoutput

    # create hdf5 to save data
    file = h5py.File(args.outputfile, "w")

    # save data to hdf5 file
    for key in data.keys():
        file.create_dataset(key, data=data[key])

    # save parameter values as attributes
    for par in set(vars(args).keys()) - {'verbose'}:
        file.attrs[par] = getattr(args, par)

    file.close()
 #   return outputdump

# Run main function
if __name__ == "__main__":
    main()

File: Julia/NetworkGame.jl
using LinearAlgebra, ArgParse, Random, Distributions, JLD2

####################################
# Structs
####################################


## organized by part of the model modified, values set via ArgParse

struct simulation_parameters
    #popgen params
    tmax::Int64
    nreps::Int64
    N::Int64
    μ::Float64
    #game params
    rounds::Int64
    fitness_benefit_scale::Float64
    b::Float64
    c::Float64
    d::Float64
    δ::Float64
    init_freqs::Vector{Float64}
    #network params
    nnet::Int64
    mutsize::Float64
    mutinitsize::Float64
    mutlink::Float64
    #file params
    filename::String
end

## smallest type necessary to play a complete round of the game 
mutable struct network
    genotype_id::Int64
    Wm
    Wb::Vector{Float64}
    InitialOffer::Float64
    CurrentOffer::Float64
end

## prevents the creation of population arrays that won't work with the shuffle reproduction method
## (those where N mod 2 != 0)
mutable struct population
    parameters::simulation_parameters
    networks::Vector{network}
    genotypes::Vector{Int64}
    fit_dict::Dict{Int64, Dict{Int64, Vector{Float64}}}
    shuffled_indices::Vector{Int64}
end

mutable struct simulation_output
    ## Fixation Stats ##
    ## time points where an allele becomes 100% of the population
    fixations::Vector{Int64}
    ## number of genotypes at each time point
    n_genotypes::Vector{Int64}

    ## simulation results ##
    ## mean values of fitness and initial offer over the sim
    w_mean_history::Vector{Float64}
    init_mean_history::Vector{Float64}

    ## Output copy of parameters
    parameters::simulation_parameters

end
####################################
# Network Game Functions
####################################



function calcOj(j::Int64, prev_out, Wm, Wb::Vector{Float64})
    ##############################
    ## Iterates a single layer of the Feed Forward network
    ##############################
    x = dot(Wm[1:j,j][1:j], prev_out[1:j]) + Wb[j]
    return 1-(exp(x*-x))
end

function iterateNetwork(input::Float64, Wm::Matrix{Float64}, Wb::Vector{Float64})
    ##############################
    ## Calculates the total output of the network,
    ## iterating over calcOj() for each layer
    ##############################

    prev_out = zeros(Float64, length(Wb))
    prev_out[1] = input
    for j in 2:length(Wb)
        prev_out[j] = calcOj(j, prev_out, Wm, Wb)
    end
    return prev_out
end

function networkGameRound(mutNet::network, resNet::network)
    ##############################
    ## Iterates above functions over a pair of networks,
    ## constitutes a single game round
    ##############################
    mutOut = last(iterateNetwork(resNet.CurrentOffer, mutNet.Wm, mutNet.Wb))
    resOut = last(iterateNetwork(mutNet.CurrentOffer, resNet.Wm, resNet.Wb))
    return [mutOut, resOut]
end

function repeatedNetworkGame(parameters::simulation_parameters, mutNet::network, resNet::network)
    ##############################
    ## Plays multiple rounds of the network game, returns differing 
    ## data types depending on whether a discount needs to be calculated
    ##############################

    mutNet.CurrentOffer = mutNet.InitialOffer
    resNet.CurrentOffer = resNet.InitialOffer

    mutHist = zeros(parameters.rounds)
    resHist = zeros(parameters.rounds)
    for i in 1:parameters.rounds
        mutNet.CurrentOffer, resNet.CurrentOffer = networkGameRound(mutNet, resNet)
        mutHist[i] = mutNet.CurrentOffer
        resHist[i] = resNet.CurrentOffer
    end
    if parameters.δ >= 0
        return [mutHist, resHist]
    elseif parameters.δ < 0
        return [mutNet.CurrentOffer, resNet.CurrentOffer]
    end
end

function fitnessOutcome(parameters::simulation_parameters,mutNet::network,resNet::network)

    ####################################
    ## calulate resident-mutant fitness matrix from contributions throughout the game history. If
    ## discount is availible, applied at rate delta going from present to the past in the 
    ## repeated cooperative investment game
    ##
    ## fitness = 1 + b * (partner) - c * (self) + d * (self) * (partner)
    ####################################
    ############################################################
    ## with discount, retrieves the full history and passes it to the
    ##  discount formula before returning final fitness value
    ############################################################
    if parameters.δ >= 0.0
        rmOut, mrOut = repeatedNetworkGame(parameters,mutNet,resNet)
        discount = exp.(-parameters.δ.*(parameters.rounds.-1 .-range(1,parameters.rounds, step = 1)))
        td = sum(discount)
        wmr = max(0, (1 + dot((parameters.b * rmOut - parameters.c * mrOut + parameters.d * rmOut.*mrOut), discount) * parameters.fitness_benefit_scale))
        wrm = max(0, (1 + dot((parameters.b * mrOut - parameters.c * rmOut + parameters.d * rmOut.*mrOut), discount)*parameters.fitness_benefit_scale))
        return [[wmr, wrm], [dot(rmOut, discount)/td, dot(mrOut, discount)/td]]

    ############################################################
    ## without discount, retrieves only the final value after all 
    ## rounds played and returns it as w based on game parameters
    ############################################################
    elseif parameters.δ < 0.0
        rmOut, mrOut = repeatedNetworkGameHistory(parameters, mutNet, resNet)
        wmr = max(0, (1 + ((parameters.b * rmOut - parameters.c * mrOut + parameters.d * rmOut.*mrOut)*parameters.fitness_benefit_scale)))
        wrm = max(0, (1 + ((parameters.b * mrOut - parameters.c * rmOut + parameters.d * rmOut.*mrOut)*parameters.fitness_benefit_scale)))
        return [[wmr, wrm], [rmOut, mrOut]]

    end
end

###############################
# Population Simulation Funcs #
###############################

function update_population!(pop::population)
    ## runs functions necessary at every timestep of the simulation
    ## updates pop struct with new partner indices and genotype ID arrays
    pop.genotypes = return_genotype_id_array(pop.networks)
    shuffle!(pop.shuffled_indices)
    update_fit_dict!(pop)
end

function return_genotype_id_array(population_array::Vector{network})
    ## WIP returns an array of the genotype inside
    genotype_array = Vector{Int64}(undef, 0)
    for individual in population_array
        append!(genotype_array, individual.genotype_id)
    end
    return genotype_array
end

function output!(t::Int64, pop::population, outputs::simulation_output)
    ## Updates output arrays
    if length(Set(pop.genotypes)) == 1
        outputs.fixations[t] = maximum(Set(pop.genotypes))
    else
        outputs.fixations[t] = 0
    end
    ## Maximum or length of the set of keys should return the largest genotype index ever present because
    ## each iteration will guarantee it shows up in fit_dict via the shuffle method
    outputs.n_genotypes[t] = length(Set(keys(pop.fit_dict)))
end
function population_construction(parameters::simulation_parameters)
    ## constructs a population array when supplied with parameters and a list of networks
    ## should default to a full array of a randomly chosen resident genotype unless
    ## instructed otherwise in params
    initialnetworks = Vector{network}(undef, length(parameters.init_freqs))
    population_array = Vector{network}(undef, 0)
    for n in 1:length(parameters.init_freqs)
        Wm = randn((parameters.nnet,parameters.nnet))

        Wb = randn(parameters.nnet)
        initOffer = (1.0 + randn())/2
        initialnetworks[n] = network(n, Wm, Wb, initOffer, initOffer)
    end
    for (net, p, gen) in zip(initialnetworks, parameters.init_freqs, 1:length(parameters.init_freqs))
        append!(population_array, repeat([net], Int64(p*parameters.N)))
    end
    return population(parameters, population_array, return_genotype_id_array(population_array), Dict{Int64, Dict{Int64, Vector{Float64}}}(), shuffle(1:parameters.N))
end

##################
# Pairwise fitness
##################
function update_fit_dict!(pop::population)
    g(x) = pop.genotypes[x]
    for (n1, n2) in zip(1:pop.parameters.N, pop.shuffled_indices)
        if g(n1) ∉ keys(pop.fit_dict)
            pop.fit_dict[g(n1)] = Dict{Int64, Vector{Float64}}()
        end
        if g(n2) ∉ keys(pop.fit_dict[g(n1)])
            fitness_outcome, raw_payoffs = fitnessOutcome(pop.parameters, pop.networks[n2], pop.networks[n1])
            pop.fit_dict[g(n1)][g(n2)] = fitness_outcome
        end
    end
end
##################
# Pairwise fitness
##################


function pairwise_fitness_calc!(pop::population)
    ## shuffles the population array, returns an array of fitness values calculated by
    ## running the fitness outcome function along both the original and shuffled array
    repro_array = Vector{Float64}(undef, pop.parameters.N)
    g(x) = pop.genotypes[x]
    for (n1,n2) in zip(1:pop.parameters.N, pop.shuffled_indices)
        repro_array[n1] = pop.fit_dict[g(n1)][g(n2)][1]./ sum(pop.fit_dict[g(n1)][g(n2)])
    end
    return repro_array
end

##################
# Reproduction function
##################

function reproduce!(pop::population)
    ## working with new arrays rather than copy of old pop to avoid in-place weirdness with shuffle()

    repro_array = pairwise_fitness_calc!(pop)
    new_genotypes = Vector{Int64}(undef, pop.parameters.N)
    new_networks = Vector{network}(undef, pop.parameters.N)
    g(x) = pop.genotypes[x]
    n(x) = pop.networks[x]
    for (res_i, mut_i) in zip(1:length(pop.networks), pop.shuffled_indices)
        if rand() <= repro_array[res_i]
            new_genotypes[res_i] = g(res_i)
            new_networks[res_i] = n(res_i)
        else
            new_genotypes[res_i] = g(mut_i)
            new_networks[res_i] = n(mut_i)
        end
    end
    pop.genotypes = new_genotypes
    pop.networks = new_networks
end

##################
#  Mutation Function 
##################

function mutate!(pop::population)
    for i in 1:length(pop.networks)
        if rand() <= pop.parameters.μ
            pop.genotypes[i] = maximum(return_genotype_id_array(pop.networks)) + 1
            mutWm = UpperTriangular(rand(Binomial(1, pop.parameters.mutlink), (pop.parameters.nnet,pop.parameters.nnet)) 
                                    .* rand(Normal(0, pop.parameters.mutsize), (pop.parameters.nnet,pop.parameters.nnet)))
            mutWb = rand(Binomial(1, pop.parameters.mutlink), pop.parameters.nnet) .* rand(Normal(0, pop.parameters.mutsize),pop.parameters.nnet)
            mutInit = rand(Normal(0, pop.parameters.mutsize))
            pop.networks[i] = network((maximum(return_genotype_id_array(pop.networks))+1),
                                        (pop.networks[i].Wm + mutWm),
                                        (pop.networks[i].Wb + mutWb),
                                        (pop.networks[i].InitialOffer + mutInit),
                                        (pop.networks[i].InitialOffer + mutInit))
        end
    end
end



#######################
# Simulation Function #
#######################

## following similar format to NetworkGame.py

function simulation(pop::population)

############
# Sim init #
############


## EG 6/4/21
## WIP Note: May need to pass a vector of initial networks + corresponding weights if want this to be 
## generalizable. Trying to do this without touching anything inside the networks struct so that I can plug JVC's
## julia network code in later.


## arrays that track population statistics
## EG 6/4/21
## WIP Note: Need to decide on output format, then create an easier to modify workflow for this.
## some kind of output struct that tracks whole sim statistics, and has vectors of timepoint statistics
## as well?

outputs = simulation_output(zeros(Int64, pop.parameters.tmax),
                            zeros(Int64, pop.parameters.tmax),
                            zeros(Float64, pop.parameters.tmax),
                            zeros(Float64, pop.parameters.tmax),
                            pop.parameters)

    ############
    # Sim Loop #
    ############
    for t in 1:pop.parameters.tmax
        # update population struct 
        update_population!(pop)

        # reproduction function / produce and save t+1 population array
        reproduce!(pop)

        # mutation function / iterates over population and mutates at chance probability μ
        mutate!(pop)

        # per-timestep counters, outputs going to disk
        output!(t, pop, outputs)

        ## should detect an error in genotype tracking
        if length(Set(keys(pop.fit_dict))) != maximum(Set(keys(pop.fit_dict)))
            print("Error in genotype tracking, dictionary of fitness values has missing genotypes")
        end
    end
## organize replicate data into appropriate data structure to be returned to main function and saved
return outputs
end

###################
#   Parameters    #
###################
function main()

    ##########
    ## Arg Parsing
    ##########

    arg_parse_settings = ArgParseSettings()
    @add_arg_table arg_parse_settings begin

        ########
        ## Population Simulation Parameters
        ########
        "--tmax"
            help = "Maximum number of timesteps"
            arg_type = Int64
            default = 500
        "--nreps"
            help = "number of replicates to run"
            arg_type = Int64
            default = 1

        "--N"   
            help = "population size"
            arg_type = Int64
            default = 100
        "--mu"
            help = "mutation probability per birth"
            arg_type = Float64
            default = 0.0

        ########
        ## Game Parameters
        ########
        "--rounds"
            help = "number of rounds the game is played between individuals"
            arg_type = Int64
            default = 5

        "--fitness_benefit_scale"
            help = "scales the fitness payout of game rounds by this amount (payoff * scale)"
            arg_type = Float64
            default = 1.0

        "--b"
            help = "payoff benefit"
            arg_type = Float64
            default = 0.0
        "--c"
            help = "payoff cost"
            arg_type = Float64 
            default = 0.0
        "--d"
            help = "payoff synergy"
            arg_type = Float64
            default = 0.0
        "--r"
            help = "relatedness coefficient"
            arg_type = Float64
            default = 0.0
        "--delta"
            help = "payoff discount, negative values use last round"
            arg_type = Float64
            default = 0.0
        "--init_freqs"
            help = "vector of initial genotype frequencies, must sum to 1"
            arg_type = Vector{Float64}
            default = [0.50,0.5]
        ########
        ## Network Parameters
        ########
        "--nnet"
            help = "network size"
            arg_type = Int64
            default = 5
        "--mutsize"
            help = "Size of mutant effects on network in Normal Dist. StdDevs"
            arg_type = Float64
            default = 0.1
        "--mutinitsize"
            help = "Size of mutant effects on initial offers in Normal Dist. StdDevs"
            arg_type = Float64
            default = 0.01
        "--mutlink"
            help = "Probability that a random edge or node be altered in a mutation event"
            arg_type = Float64
            default = 0.5
        ########
        ## File Parameters
        ########
        "--filename"
            help = "File to save outputs to"
            arg_type = String
            default = "NetworkGameOutput.jld2"
    end
    
    ##passing command line arguments to simulation
    parsed_args = parse_args(ARGS, arg_parse_settings)
    parameters = simulation_parameters(parsed_args["tmax"], parsed_args["nreps"], parsed_args["N"], parsed_args["mu"],
                                        parsed_args["rounds"], parsed_args["fitness_benefit_scale"], parsed_args["b"], 
                                        parsed_args["c"], parsed_args["d"], parsed_args["delta"], parsed_args["init_freqs"], 
                                        parsed_args["nnet"], parsed_args["mutsize"], parsed_args["mutinitsize"], parsed_args["mutlink"])

    ## Necessary sanity checks for params
    if mod(parameters.N, 2) != 0
        print("Please supply an even value of N!")
    end

    ##############
    ## Test Values for comparing to python (or other) implementation
    ##############

    # nnet = 2
    # initWm = transpose(reshape([0.71824181,2.02987316,-0.42858626,0.6634413],2,2))
    # initWb = [-0.66332791,1.00430577]
    # init = 0.1
    # InitialNetwork = network(initWm, initWb, init, init)
    
    # Wm = [0.0 0.0; 0.0 0.0]
    # Wb = [0.0, 0.0]
    # Wm2 = [0.71824181 2.02987316; -0.42858626 0.6634413]
    # Wb2 = [-0.66332791,1.00430577]
    # init = .1
    # initnet1 = network(1, Wm2, Wb2, init, init)
    # initnet2 = network(2, Wm2, Wb2, init, init)
    # (fitnessOutcome(parameters, initnet1, initnet2))
    ##################################
    #Generation of Random Initial Network
    ##################################
    ## generates random networks based on simulation parameters
    init_pop = population_construction(parameters)

    ###################
    # Simulation call #
    ###################
    for rep in 1:parameters.nreps
        replicate_output = simulation(init_pop)
    end


    ###################
    #   Data Output   #
    ###################

    ## sim data will be stored in previous section and saved to disk here. 


    ###################
    #     Visuals     #
    ###################
    
    ## should be it's own script eventually. Anything necessary for debugging should be done here
    ## also, test out passing the files to some graphics function at some point if it becomes an issue


end


main()

